KAFKA-15931: Reopen TransactionIndex if channel is closed

jeqo opened the pull request:
Cached TransactionIndex may get closed if interrupted, causing following calls to always fail with ClosedChannelException, and forcing process to be restarted. In order to avoid this issue, a new method is exposed by TransactionIndex to validate state of channel; and index is reopened if closed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jeqo committed:
KAFKA-15931: Reopen TransactionIndex if channel is closed

Cached TransactionIndex may get closed if interrupted, causing following calls to always fail with ClosedChannelException, and forcing process to be restarted.
In order to avoid this issue, a new method is exposed by TransactionIndex to validate state of channel; and index is reopened if closed.

nikramakrishnan commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -522,6 +522,18 @@ public TimeIndex timeIndex() {
 
         // Visible for testing
         public TransactionIndex txnIndex() {
+            if (txnIndex.isClosed()) {
```
Nit: Maybe it makes sense to add a debug log here to indicate that the index was closed and is being reopened? Might be useful to debug any performance issues.

nikramakrishnan reviewed (approved):
Thanks for the fix, @jeqo! These changes make sense to me on first pass.

nikramakrishnan commented:
Bump! @satishd @kamalcph can we get this review going? Thanks!

kamalcph commented:
Thanks for the patch @jeqo! 

I was able to reproduce the exception mentioned in the ticket by calling close in-middle of collecting the aborted transactions. The fix doesn't resolve the issue. Could you please write a unit test to ensure that the issue gets resolved?

https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java#L177

```
org.apache.kafka.common.KafkaException: Failed read position from the transaction index /var/folders/bq/w6tnkbq964q8sqpvj4fbmjr40000gq/T/kafka9502115127128961380.tmp
	at org.apache.kafka.storage.internals.log.TransactionIndex$1.hasNext(TransactionIndex.java:245)
	at org.apache.kafka.storage.internals.log.TransactionIndex.collectAbortedTxns(TransactionIndex.java:176)
	at kafka.log.TransactionIndexTest.testCollectAbortedTxnsDuringClose(TransactionIndexTest.scala:210)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        ...
Caused by: java.nio.channels.ClosedChannelException
	at java.base/sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:150)
	at java.base/sun.nio.ch.FileChannelImpl.position(FileChannelImpl.java:325)
	at org.apache.kafka.storage.internals.log.TransactionIndex$1.hasNext(TransactionIndex.java:243)
	... 86 more

```

showuon commented:
@jeqo , could you address @kamalcph 's comment above? Thanks.

jeqo committed:
fixup! KAFKA-15931: Reopen TransactionIndex if channel is closed

Refactor search for aborted txns

jeqo commented:
@kamalcph good catch! You're right. I have refactored the code a bit further to reopen the txn index properly. PTAL

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -633,6 +657,10 @@ public String toString() {
         }
+
+        public boolean hasTxnIndex() {
+            return txnIndex.file().length() == 0;
```
is this check valid? should we have to invert it?

jeqo replied:
my mistake, fixing it

kamalcph reviewed (commented):
There are multiple RemoteIndexCache test failures in the latest [run](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15241/2/tests). PTAL.

kamalcph commented:
@jeqo 

Any updates on this PR?

jeqo committed:
refactor: collecting aborted transactions from remote files

Entry always contains txn index. If not existing on source, then it's an empty file on the cached entry.
Then, with this assumption, the logic to collect aborted transactions can be simplified.

jeqo committed:
refactor: allow txn index to be read after closed

It is responsibility of the client to close the transaction index, but the transaction index should allow to read values even after closing.
The channel is an internal detail and should not affect reading values. If the file is available, then it should be read.
To encapsulate this dependency between file and buffer an internal abstraction TransactionIndexFile is introduced to handle the coordination between these two values.

jeqo committed:
Revert "KAFKA-15931: Reopen TransactionIndex if channel is closed"

This reverts commit 250c2c2222ce243313613daebd708dfa7ff770d2.

jeqo committed:
Revert "fixup! KAFKA-15931: Reopen TransactionIndex if channel is closed"

This reverts commit ed29d4da4c7ec7b60d30a692048afe5a215822fc.

kamalcph commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -1622,16 +1622,12 @@ private void collectAbortedTransactions(long startOffset,
+            TxnIndexSearchResult searchResult = transactionIndex.collectAbortedTxns(startOffset, upperBoundOffset);
+            accumulator.accept(searchResult.abortedTransactions);
+            // Return immediately when the search result is complete, it does not need to go through local log segments.
+            if (searchResult.isComplete) return;
```
Thanks for the nice refactor!

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java:
```diff
@@ -190,75 +175,145 @@ public void sanityCheck() {
-                    return channel.position() - position.value >= AbortedTxn.TOTAL_SIZE;
-                } catch (IOException e) {
-                    throw new KafkaException("Failed read position from the transaction index " + file.getAbsolutePath(), e);
+    public static class TransactionIndexFile {
```
Can we make this class as `private`? We can expose the `channel()` with package-private modifier for the tests in TransactionIndex.

jeqo replied:
Seems that by making it private, it can only be accessed from TransactionIndex class but not tests or others on the same module. 
Switching access to package-private and adding comment on visibility on last commit.

kamalcph reviewed (commented):
Thanks for the patch, left one minor comment to address.

jeqo committed:
fixup! refactor: allow txn index to be read after closed

Reduce access to TransactionIndexFile.

showuon reviewed (commented):
@jeqo , thanks for the PR. But I think the root cause of this problem is the race condition between one thread trying to close the channel and the other thread trying to read the index. It can be reproduced with the following patch:

```
git diff
diff --git a/storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java b/storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java
index 10d1449cdb..b39317ef19 100644
--- a/storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java
+++ b/storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java
@@ -112,7 +112,7 @@ public class TransactionIndex implements Closeable {
         FileChannel channel = channelOrNull();
         if (channel != null)
             channel.close();
-        maybeChannel = Optional.empty();
+        // maybeChannel = Optional.empty();
     }
 
     /**
diff --git a/storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java b/storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java
index 7a75ab0cce..3598b00197 100644
--- a/storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java
+++ b/storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java
@@ -202,4 +202,11 @@ public class TransactionIndexTest {
         index.deleteIfExists();
         assertFalse(file.exists());
     }
+
+    @Test
+    public void lukeTest() throws IOException {
+        assertTrue(file.exists());
+        index.close();
+        index.collectAbortedTxns(0L, 100L);
+    }
```

```
Gradle Test Run :storage:test > Gradle Test Executor 17 > TransactionIndexTest > lukeTest() FAILED
    org.apache.kafka.common.KafkaException: Failed read position from the transaction index /var/folders/qx/r9z7qcz96_nbbk3b38bhl32m0000gn/T/kafka1652929925255619923.tmp
        at app//org.apache.kafka.storage.internals.log.TransactionIndex$1.hasNext(TransactionIndex.java:235)
        at app//org.apache.kafka.storage.internals.log.TransactionIndex.collectAbortedTxns(TransactionIndex.java:171)
        at app//org.apache.kafka.storage.internals.log.TransactionIndexTest.lukeTest(TransactionIndexTest.java:210)

        Caused by:
        java.nio.channels.ClosedChannelException
            at java.base/sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:159)
            at java.base/sun.nio.ch.FileChannelImpl.position(FileChannelImpl.java:352)
            at org.apache.kafka.storage.internals.log.TransactionIndex$1.hasNext(TransactionIndex.java:233)
            ... 2 more

```

That is, when one thread completed `channel.close()`, but `maybeChannel` is still non-null, at the moment, the other thread will get the channel close exception while reading the channel. 

To fix the problem, we should using a lock to protect this race condition, instead of re-opening the channel, because the channel close must have a reason. If we can prevent the race condition, the read after close will not throw exception, but just get empty result. If the empty result is not expected, we should, then, investigate why the channel is closed earlier. 

Does that make sense?

jeqo committed:
refactor: switch from File to Path when possible

Simplify code by using Path directly on TransactionIndexFile

jeqo committed:
refactor: move channel access into TransactionIndexFile

Reduce access to channel by exposing read/write/position on the TransactionIndexFile.

jeqo commented:
Thanks @showuon. That's a fair point. 
The main issue this PR was trying to address was the fact that the cached Entry was left with a TransactionIndex closed, hence the reopening solution.
I wonder if the race condition may have been already fixed while this PR has been in review--because there is no obvious call to `.close()` where we can follow an investigation.
For instance, isn't https://github.com/apache/kafka/commit/88d2c4460a1c8c8cf5dbcc9edb43f42fe898ca00 actually solving the race condition? Myguess is that by using the "wrong" lock the close was not blocking reads from the cache, hence the bug. And seems this commit already fixes this on 3.7 -- @divijvaidya wonder if you could confirm?

In that case, I could remove the channel reopening and leave this as a refactoring PR if that's useful. Thanks!

jeqo committed:
fixup! refactor: allow txn index to be read after closed

Test `ClosedByInterruptException` exception and use synchronized methods
to access channel;

jeqo commented:
We have received feedback of this issue persisting on 3.8, so I guess we could have another look at this PR.

@showuon I have made the TransactionIndexFile channel access methods synchronized. Given that this is not an index expected to be concurrently accessed by many threads, this approach should be fine.
Also, got to know[1] that if the thread accessing the channel is interrupted, the channel will get closed implicitly. I think this is the main cause, as interruption are the way to handle delayed fetch operations. I have added a test to validate the channel gets reopen after interruption. PTAL.

[1]: https://stackoverflow.com/questions/1161297/why-are-we-getting-closedbyinterruptexception-from-filechannel-map-in-java-1-6/1161455#1161455

mjd95 commented:
We were the ones discussing with @jeqo - the "caching closed channels" issue was happening regularly for us on 3.8 in production, the thread doing a remote read was interrupted while iterating through the transaction index, we get a `ClosedByInterruptException` on some transaction index file channel, but the closed channel remains in the cache. The only way to mitigate was restarting the broker.

We were able to reproduce by setting a low `remote.fetch.max.wait.ms` and setting a small segment size in order to generate many transaction index files.

We tested that this PR fixes our repro and cherry-picked it onto our production release, we haven't seen the issue since then.

We haven't seen the "race during channel close" issue (which is now also handled in this PR) in production.

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java:
```diff
@@ -238,35 +207,137 @@ private Iterable<AbortedTxnWithPosition> iterable(Supplier<ByteBuffer> allocate)
+            Utils.writeFully(channel(), buffer);
+        }
+
+        void read(ByteBuffer buffer, int position) throws IOException {
```
can we rename this method to `readFully` as it doesn't read byte-by-byte?

kamalcph commented:
The patch LGTM. It needs second set of eyes as we are re-opening a closed index. cc @showuon @satishd PTAL.

jeqo committed:
fixup! refactor: allow txn index to be read after closed

Apply suggestion https://github.com/apache/kafka/pull/15241#discussion_r1835574783

showuon reviewed (approved):
LGTM!

showuon commented:
@jeqo , could you try to rebase to the latest trunk? I'd like to verify the change doesn't break the build. Thanks.

jeqo commented:
@showuon thanks! build looks good after rebase 👍🏽

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/log/TransactionIndex.java:
```diff
@@ -238,35 +207,137 @@ private Iterable<AbortedTxnWithPosition> iterable(Supplier<ByteBuffer> allocate)
+            channel.position(channel.size());
+        }
+
+        synchronized void updateParentDir(Path parentDir) {
```
Having so many synchronized methods doesn't seem to follow the pattern we have used in the log layer. @junrao Thoughts?

junrao replied:
The solution is not ideal. When the channel is closed, it's possible that a LocalLog.read() is also reading from the channel and it will hit `ClosedByInterruptException`. Since this is an IOException, the `maybeHandleIOException()` logic will treat it as a failed disk and kill the broker (if there is only one disk).

I think it's better to change the logic in RemoteLogReader such that it never interrupts the thread. Note that we made other threads like ReplicaFetcherThread and CleanerThread non-interruptible for the same reason. So, this approach will be consistent with the existing ones.

Specially, we could change the following code to cancel without interruption.

`    val cancelled = remoteFetchTask.cancel(true)`

We can then pass in a cancelled flag in RemoteLogReader. The flag will be set when the task is cancelled and `RemoteLogManager.read()` can check this flag to abort the operation early if needed. This way, we can avoid the risk of false disk failure and the overhead of reopening the file channel for txn index.

ijuma replied:
That makes sense to me. @jeqo Will you please take a look at this? Perhaps we should revert this change and start again with the approach @junrao suggested.

jeqo replied:
Sure, will have a look

jeqo replied:
@junrao Thanks for the suggestion and making the connection between the DelayedRemoteFetch and the RemoteLogReader getting interrupted. I was missing that.
I have a draft for the suggestion here: https://github.com/apache/kafka/pull/19197/commits/0f1332514234536bece94afc57f18f024174a9c6, let me know if it goes along the same lines.

junrao replied:
Thanks for the draft, @jeqo. Left a few comments in https://github.com/apache/kafka/pull/19197.

junrao reviewed (commented):
@jeqo : Thanks for the PR and sorry for the late review. Left a comment below.
